File: adk-go/src/application/tools/validators.go
package tools

import (
	"regexp"
	"strconv"
)

type ValidationResult struct {
	Valid   bool   `json:"valid"`
	Message string `json:"message"`
}

func ValidateNip(nip string) ValidationResult {
	re := regexp.MustCompile(`\D`)
	digits := re.ReplaceAllString(nip, "")

	if len(digits) != 10 {
		return ValidationResult{Valid: false, Message: "NIP musi mieć 10 cyfr."}
	}

	weights :=[]int{6, 5, 7, 2, 3, 4, 5, 6, 7}
	checksum := 0
	for i := 0; i < 9; i++ {
		d, _ := strconv.Atoi(string(digits[i]))
		checksum += d * weights[i]
	}
	checksum = checksum % 11

	lastDigit, _ := strconv.Atoi(string(digits[9]))
	if checksum == lastDigit {
		return ValidationResult{Valid: true, Message: "NIP jest poprawny."}
	}
	return ValidationResult{Valid: false, Message: "NIP ma błędną sumę kontrolną."}
}

File: baseline/src/application/tools/validators.py
import re

def validate_nip(nip: str) -> dict:
    """Sprawdza poprawność numeru NIP (polska suma kontrolna).

    Args:
        nip: Numer NIP jako string (może zawierać myślniki/spacje).

    Returns:
        Słownik z polem 'valid' (bool) i 'message' (str).
    """
    digits = re.sub(r"\D", "", nip)
    if len(digits) != 10: return {"valid": False, "message": f"NIP musi mieć 10 cyfr, podano {len(digits)}."}
    weights =[6, 5, 7, 2, 3, 4, 5, 6, 7]
    checksum = sum(int(digits[i]) * weights[i] for i in range(9)) % 11
    if checksum == int(digits[9]): return {"valid": True, "message": "NIP jest poprawny."}
    return {"valid": False, "message": "NIP ma błędną sumę kontrolną."}